# Instructions

- This homework assignment is worth **51** points. 
- Please strive for clarity and organization.
- **AI Usage**:
   - AI use is strictly **prohibited** for conceptual exercises.
   - For applied (Python) exercises, you may only use AI for **debugging**. Do not use AI to generate answers.
- **Due Date: September 18, 2026, by 11:59 PM.**

# Exercise 1 

In this exercise, we will keep practicing basic operations with `pandas`. 

### Exercise 1(a) (2 points)

Import the `pandas` library.

In [1]:
import pandas as pd

### Exercise 1(b) (3 points)

Create a data-frame called `euro12` using the data from `Euro_2012_stats_TEAM.csv`.

In [6]:
euro12 = pd.read_csv('Euro_2012_stats_TEAM.csv')# Your answer here

### Exercise 1(c) (2 points)

Compute the average number of `Goals`.

In [8]:
euro12['Goals'].mean()

np.float64(4.75)

### Exercise 1(d) (2 points)

How many teams participated in the Euro 2012?

In [9]:
euro12['Team'].nunique()

16

### Exercise 1(e) (3 points)

Create a data-frame called `discipline` using `Team`, `Yellow Cards`, and `Red Cards` from the `euro12` data-frame.

In [12]:
discipline = euro12[['Team', 'Yellow Cards', 'Red Cards']] 

### Exercise 1(f) (3 points)

Sort the `discipline` data-frame based on `Red Cards`.

In [13]:
discipline.sort_values(by='Red Cards') 

,Team,Yellow Cards,Red Cards
0,Croatia,9,0
1,Czech Republic,7,0
2,Denmark,4,0
3,England,5,0
4,France,6,0
5,Germany,4,0
7,Italy,16,0
8,Netherlands,5,0
10,Portugal,12,0
12,Russia,6,0


### Exercise 1(g) (3 points)

Using the `discipline`, compute the average number of `Yellow Cards`.

In [14]:
discipline['Yellow Cards'].mean()

np.float64(7.4375)

### Exercise 1(h) (3 points)

Using the `euro12` data-frame, report the team that scored more than 6 goals.

In [15]:
euro12[euro12['Goals'] > 6] ['Team'] 

5     Germany
13      Spain
Name: Team, dtype: object

### Exercise 1(i) (3 points)

Using the `euro12` data-frame, report the teams that start with `S`.

In [16]:
euro12[euro12['Team'].str.startswith('S')]['Team'] 

13     Spain
14    Sweden
Name: Team, dtype: object

# Exercise 2 (10 points)

In this exercise, you will use the `Raisin_Dataset.csv`. For more details about the dataset, see this [link](https://www.kaggle.com/datasets/nimapourmoradi/raisin-binary-classification/data).
Here's a practical exercise highlighting modularity's importance in data science by refactoring non-modular code into modular, reusable components. You are given a simple data science task: load a dataset, clean it, perform some feature engineering, train a machine learning model, and evaluate the results. However, the initial code is not modular—it is repetitive and hard to maintain.

In [13]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

# Load dataset
df = pd.read_csv('Raisin_Dataset.csv')

# # Clean dataset
df = df.dropna()
df['Area/Perimeter'] = df['Area'] / df['Perimeter']
df['sqrt_area'] = np.sqrt(df['Area'])
df['ext/ecc'] = df['Extent'] / df['Eccentricity']
df['convA/Area'] = df['ConvexArea'] / df['Area']
df['Class'] = df['Class'].map({'Besni':0, 'Kecimen':1})

# Split dataset into training and testing
X = df.drop(columns='Class', axis=1)
y = df['Class']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Train model
model = RandomForestClassifier()
model.fit(X_train, y_train)

# Make predictions
y_pred = model.predict(X_test)

# Evaluate model
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy}")

Accuracy: 0.85


This script works, but it's not modular. If you needed to modify the data-cleaning process or change the model, you’d have to go through the whole script and make changes in several places. Now, refactor the code into modular, reusable components by turning each logical part into functions. This modular design will make it easier to maintain, test, and reuse for different datasets or models.

In [22]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split 
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score

def load_data(path):
    df = pd.read_csv(path)
    return df

def clean_data(df):
    df = df.dropna()
    return df

def engineer_features(df):
    df['Area/Perimeter'] = df['Area'] / df['Perimeter']
    df['sqrt_area'] = np.sqrt(df['Area'])
    df['ext/ecc'] = df['Extent'] / df['Eccentricity']
    df['convA/Area'] = df['ConvexArea'] / df['Area']
    df['Class'] = df['Class'].map({'Besni':0, 'Kecimen':1})
    return df

def split_data(df):
    X = df.drop(columns='Class', axis=1)
    y = df['Class']
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
    return X_train, X_test, y_train, y_test

def train_model(X_train, y_train):
    model = RandomForestClassifier()
    model.fit(X_train, y_train)
    return model

def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)
    accuracy = accuracy_score(y_test, y_pred)
    return accuracy

if __name__ == "__main__":
    # Load and clean data
    df = load_data('Raisin_Dataset.csv')  
    df = clean_data(df)
    
    # Feature engineering
    df = engineer_features(df)
    
    # Split data
    X_train, X_test, y_train, y_test = split_data(df)
    
    # Train model
    model = train_model(X_train, y_train)
    
    # Evaluate model
    accuracy = evaluate_model(model, X_test, y_test)
    print(f"Accuracy: {accuracy}")

Accuracy: 0.8555555555555555


# Exercise 3 (8 points)

In this exercise, you will time all the component of your answer from Exercise 2. Report the times. 

1. Which part of the code is the bottleneck?
2. What would happen if the dataset grew in size? Would certain parts of the code scale efficiently, or would they slow down significantly?


In [23]:
import time

start = time.perf_counter()
df = load_data('Raisin_Dataset.csv') 
end = time.perf_counter()
print("Load data:", end - start) 

start = time.perf_counter()
df = clean_data(df)
end = time.perf_counter()
print("Clean data:", end - start)

start = time.perf_counter()
df = engineer_features(df)
end = time.perf_counter()
print("Feature engineering:", end - start)

Load data: 0.0064794999998412095
Clean data: 0.0009152089996860013
Feature engineering: 0.002111417000378424


In [24]:
start = time.perf_counter()
X_train, X_test, y_train, y_test = split_data(df)
end = time.perf_counter()
print("Split data:", end - start)

start = time.perf_counter()
model = train_model(X_train, y_train)
end = time.perf_counter()
print("Train model:", end - start)

start = time.perf_counter()
accuracy = evaluate_model(model, X_test, y_test)
end = time.perf_counter()
print("Evaluate model:", end - start)

Split data: 0.0033786250000957807
Train model: 0.08872408299976087
Evaluate model: 0.003387875000044005


1. The bottleneck is training the model because it took approximately 0.089 seconds, which was much longer than the other pats of the code.

2. I would expect the code to take longer to run. Training the Random Forest model would likely slow down because it has to process more data and build multiple decision trees. I would also expect loading, cleaning, and feature engineering would take longer with more data,  but based on the results, model training would be my main concern.

# Exercise 4 (9 points)

The following questions are conceptual. Answer them in your own words, in complete sentences. **No AI use is permitted for these questions.**


### Exercise 4(a) (3 points)

In your own words, explain the *Don't Repeat Yourself* (DRY) principle. Using the code from Exercise 2 as your example, identify one place where repetition (or near-repetition) occurs and explain what problem would arise later if that repeated logic had to change.


The DRY principle means avoiding repeated code or logic when you can use somethings once and reuse it. During Exercise 2, I have a similar feature-engineering calculation where I create a new column by dividing two existing columns. If I needed to change how these calculations were done, I would have to change each similar line separatly, whcih could the code a bit harder to maintain and could potentially lead to mistakes.

### Exercise 4(b) (3 points)

A classmate says: "My script runs and gives the right answer, so the code is good." Using at least three of the five principles of good code (simplicity, modularity, readability, performance, robustness), explain why producing the correct output is not enough to call code good.


Just beacause the script runs and produces the correct output, does not necessarily mean the good is good. Good code should be simple, readable, modular, and robust. During my Exercise two, my code gives the correct accuracy of 0.8555555555555555, but it is not very modular becuase the data cleaning, feature engineering, and modle training are all in one block. Once we begin breaking these steps into functions, the code would be much easier to read, resue, test, and change later (if needed).

### Exercise 4(c) (3 points)

Suppose the `Raisin_Dataset.csv` file used in Exercise 2 is missing a few values in the `Area` column, and later the file is renamed. Explain what *robustness* means in this context, and describe two specific things you would add to your Exercise 2 functions so that the program does not fail silently. In your answer, explain the difference between an inline comment and a docstring, and state which one you would use to document the behavior you added.


Robustness means that a program can handlde unexpected problems wihtout failing. In Exercise 2, I would add a check for missing values in the Area column and error handling if the CSV file is missing or renamed. I would also take note with how the program handles missing data and file errors, as it has happened countless times with other exercises.